### PR TITLE
Move from reliance on p parameter to using PATH_INFO

### PR DESCRIPTION
**Pull request by initvector**
Reliance on the p parameter to determine the requested path causes conflicts with vBulletin routes.  This fix allows usage of the PATH_INFO server variable, if available, to determine the requested path information.

Also some caching goodness added to .htaccess for the following extensions:

* css
* js
* woff
* ttf
* eot
* svg
* png
* gif
* jpeg
* jpg
* ico
* swf

### CODE REVIEW

**Comment by tburry** [comment]
@linc, Can you see if anyone in the community is willing to test this? OSS is where this will see issues with the various environments.

File: library/core/class.request.php [hunk 1]
```diff
@@ -435,7 +435,9 @@ protected function _LoadEnvironment() {
             SafeParseStr($Get, $Get, $Original);
          }
 
-         if (isset($Get['_p'])) {
+         if (isset($_SERVER['X_REWRITE']) && isset($_SERVER['PATH_INFO'])) {
```
Comment (tburry) [inline library/core/class.request.php]: Is `isset($_SERVER['PATH_INFO'])` always set? I'm thinking of the case where you are on the server root. Maybe we can just check for `isset($_SERVER['X_REWRITE'])`.

**Comment by initvector** [comment]
It's part of the CGI spec, so I'm *pretty sure* it'll always be set.  I'll be more confident in that, once we're able to confirm functionality in less-common setups.

**Comment by tburry** [comment]
One thing this PR will break is embed.js because it always sends index.php?p=. You will probably have to do the following:

1. Modify embed.js to use the proper path.
2. Make this work for sites that have X_REWRITE, but call the site with index.php?p=.

Let me know what you think here.

File: .htaccess [hunk 2]
```diff
@@ -1,12 +1,29 @@
...
+
+    # 301 redirect urls that start with index.php
+    #RewriteCond %{REQUEST_METHOD} GET [NC]
+    #RewriteCond %{REQUEST_URI} ^(.*?)/index\.php(.*)$
+    #RewriteRule ^index\.php /%1%2 [QSA,R,L]
+</IfModule>
+
+<IfModule mod_headers.c>
+   <FilesMatch "\.(css|js|woff|ttf|eot|svg|png|gif|jpeg|jpg|ico|swf)$">
+      Header set Cache-Control "max-age=315360000"
```
Comment (tburry) [inline .htaccess]: Don't cache `embed.js` or `vanilla.embed.js`.

File: library/core/class.request.php [hunk 3]
```diff
@@ -435,7 +435,9 @@ protected function _LoadEnvironment() {
             SafeParseStr($Get, $Get, $Original);
          }
 
-         if (isset($Get['_p'])) {
+         if (isset($_SERVER['X_REWRITE'])) {
+            $Path = $_SERVER['PATH_INFO'];
+         } elseif (isset($Get['_p'])) {
```
Comment (tburry) [inline library/core/class.request.php]: Remove support for `_p`.

**Comment by tburry** [comment]
I think this PR is ready to go at this point, but we need some help with apache testing. Also @linc can maybe give it a smoke test locally and let us know that we are past OSS 2.2 before this is merged.

File: .htaccess [hunk 4]
```diff
@@ -1,12 +1,32 @@
...
+    RewriteEngine On
+    # Certain hosts may require the following line.
+    # If vanilla is in a subfolder then you need to specify it after the /.
+    # (ex. You put Vanilla in /forum so change the next line to: RewriteBase /forum)
+    # RewriteBase /
+
+    RewriteCond %{QUERY_STRING} ^p=/?([^&]+)(&(.+))?$
+    RewriteRule ^index\.php index.php/%1?%3 [E=X_REWRITE:1,L]
+
+    # The basic rewrtie rule.
```
Comment (linc) [inline .htaccess]: rewrtie

**Pull request merged**